{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import toolkit as tk\n",
    "\n",
    "# Specify the file to use for analysis \n",
    "# This should be a .txt file generated by preprocessing a fastq file (see the DelSequencingNotebook for performing preprocessing)\n",
    "PROC_FASTQ_FILE = \"ExampleData_Preprocessed.txt\"\n",
    "\n",
    "# The number of sequences reads to perform analysis with (500 is typically sufficient)\n",
    "n_sequences=500\n",
    "\n",
    "\n",
    "#############################################################\n",
    "## Code below this point does not require any modification ##\n",
    "#############################################################\n",
    "\n",
    "seq_list=[]\n",
    "seq_counter=0\n",
    "mapping_dict={'A':1,'T':2,'C':3,'G':4}\n",
    "\n",
    "\n",
    "with open(PROC_FASTQ_FILE, 'r') as seq_file:\n",
    "   #extract the first n_sequences into  a list\n",
    "    for record in tqdm(seq_file):\n",
    "        seq_counter+=1\n",
    "        seq_list.append(record)\n",
    "        if seq_counter==n_sequences:\n",
    "            break\n",
    "\n",
    "# Create a character matrix from the extracted sequences\n",
    "# A mapping dict is used to map bp strings to integers so they can be visualized using a heatmap\n",
    "int_matrix = [[mapping_dict[str(seq[idx])] for seq in seq_list] for idx in range(len(seq_list[0][:-1]))]\n",
    "int_df = pd.DataFrame(int_matrix)\n",
    "int_df=int_df.transpose()\n",
    "int_df.columns+=1 #Shift the column vals so they start at 1 (for easier interpretation)\n",
    "\n",
    "# Calculate the column-wise variance for the int_df\n",
    "# Areas with higher variance should correspond to encoding regions\n",
    "seq_variance=tk.calc_variance(int_df)\n",
    "\n",
    "# Create a heatmap\n",
    "plt.figure(figsize=(40, 20))\n",
    "ax1 = plt.subplot(2, 1, 1)\n",
    "sns.barplot(x=range(1, len(seq_variance) + 1), y=seq_variance, ax=ax1) #x vals are specifically started at 1\n",
    "\n",
    "# Create a bar plot\n",
    "ax2 = plt.subplot(2, 1, 2)\n",
    "sns.heatmap(int_df,  cmap='YlGnBu', ax=ax2, cbar=False)\n",
    "\n",
    "# Plot the data\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output the sequences used in the SeqView to SeqViewSequences.txt <br>\n",
    "* (this is useful if your PC has trouble opening the full prepocessed .txt file due to filesize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"SeqViewSequences.txt\", \"w+\") as output_file:\n",
    "        with open(PROC_FASTQ_FILE, 'r') as seq_file:\n",
    "            seq_lines=seq_file.readlines()[:n_sequences] \n",
    "            for seq in seq_lines:\n",
    "                output_file.write(seq)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
